### PR TITLE
gh-93691: Compiler's code-gen passes location around instead of holding it on the global compiler state

### PR DESCRIPTION
**Pull request by iritkatriel**
This implements https://github.com/python/cpython/issues/93691#issuecomment-1153797347. 

The line numbers are almost identical to those we calculated before (see the individual commits if you're interested in the process I followed).   There were a few place (around annotations, for instance) where the line number was not set so it remained whatever it was before. I didn't bother going through hoops to make this backwards compatible because it's obviously wrong. I just take the line number from the current AST node. 

In a few places (like pattern matching) I pass around a pointer to the location which can be updated deep in the recursion. This is not how it should be, but it's how it was until now (the global was updated in the recursion and impacted further nodes). So I used pointers to emulate this and get similar results, but my intention is to get rid of these pointers in the future, and just take locations from the relevant, current, AST node.

* Issue: gh-93691

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
Going to retrigger buildbots after merging the refleak fix.

**Comment by markshannon** [comment]
I notice that some function take a location, and others take a pointer to a location. Why?
A typedef might help readability as well.
```
typedef struct location { ... } Location;
```

**Comment by iritkatriel** [comment]
> I notice that some function take a location, and others take a pointer to a location. Why? 

See above (https://github.com/python/cpython/pull/98001#issue-1400409860).

> A typedef might help readability as well.
> 
> ```
> typedef struct location { ... } Location;
> ```

Yeah could do.

File: Python/compile.c [hunk 1]
```diff
@@ -3186,79 +3194,95 @@ compiler_while(struct compiler *c, stmt_ty s)
...
-    ADDOP_JUMP(c, JUMP, loop->fb_block);
+    ADDOP_JUMP(c, loc, JUMP, loop->fb_block);
     return 1;
 }
 
 
+static location
+last_location_in_body(asdl_stmt_seq *stmts)
+{
+    for (int i = asdl_seq_LEN(stmts) - 1; i >= 0; i++) {
```
Comment (markshannon) [inline Python/compile.c]: WIndows warning. Use `Py_ssize_t` instead of `int`.

File: Python/compile.c [hunk 2]
```diff
@@ -3186,79 +3194,95 @@ compiler_while(struct compiler *c, stmt_ty s)
...
+        return compiler_error(c, loc, "'continue' not properly in loop");
     }
-    ADDOP_JUMP(c, JUMP, loop->fb_block);
+    ADDOP_JUMP(c, loc, JUMP, loop->fb_block);
     return 1;
 }
 
 
+static location
+last_location_in_body(asdl_stmt_seq *stmts)
```
Comment (markshannon) [inline Python/compile.c]: This isn't necessarily the last location, but the location of the last executing statement.
I can't think of a case where those would be different, but might be worth a comment.

File: Python/compile.c [hunk 3]
```diff
@@ -3309,25 +3335,29 @@ compiler_try_finally(struct compiler *c, stmt_ty s)
...
+    ADDOP_JUMP(c, loc, SETUP_CLEANUP, cleanup);
+    ADDOP(c, loc, PUSH_EXC_INFO);
+    if (!compiler_push_fblock(c, loc, FINALLY_END, end, NO_LABEL, NULL))
         return 0;
     VISIT_SEQ(c, stmt, s->v.Try.finalbody);
+    loc = last_location_in_body(s->v.Try.finalbody);
     compiler_pop_fblock(c, FINALLY_END, end);
-    ADDOP_I(c, RERAISE, 0);
+
+    ADDOP_I(c, loc, RERAISE, 0);  // CHANGED
```
Comment (markshannon) [inline Python/compile.c]: What is the meaning of the comment? (there are a few of these)

File: Python/compile.c [hunk 4]
```diff
@@ -3473,62 +3511,64 @@ compiler_try_except(struct compiler *c, stmt_ty s)
...
-            ADDOP(c, POP_TOP); /* exc_value */
+            ADDOP(c, loc, POP_TOP); /* exc_value */
 
             USE_LABEL(c, cleanup_body);
-            if (!compiler_push_fblock(c, HANDLER_CLEANUP, cleanup_body, NO_LABEL, NULL))
+            if (!compiler_push_fblock(c, loc, HANDLER_CLEANUP, cleanup_body, NO_LABEL, NULL))
                 return 0;
             VISIT_SEQ(c, stmt, handler->v.ExceptHandler.body);
             compiler_pop_fblock(c, HANDLER_CLEANUP, cleanup_body);
             UNSET_LOC(c);
```
Comment (markshannon) [inline Python/compile.c:3557]: Is this needed any more?

File: Python/compile.c [hunk 5]
```diff
@@ -4015,8 +4066,10 @@ compiler_visit_stmt(struct compiler *c, stmt_ty s)
                 n++;
             }
         }
-        ADDOP_I(c, RAISE_VARARGS, (int)n);
+        location loc = LOC(s);
```
Comment (markshannon) [inline Python/compile.c:4069]: There's a few cases of this pattern:
```C
    location loc = LOC(s);
    ADDOP...(..., loc, ...);
```
Any reason not to shorten this to:
```C
    ADDOP...(..., LOC(S), ...);
```
?

**Review by markshannon — CHANGES_REQUESTED** [review]
Nice improvement.

A few comments, but no major issues.

File: Python/compile.c (hunk 1 above)
Comment (iritkatriel) [reply Python/compile.c]: Why is this? There are a few other for loops with int variable in this file, I don't know if they need to change too.

File: Python/compile.c (hunk 2 above)
Comment (iritkatriel) [reply Python/compile.c]: I wonder if there's a bug here (not a new one, in main too) when there is dead code at the end of a finally block.

File: Python/compile.c (hunk 4 above)
Comment (iritkatriel) [reply Python/compile.c:3557]: It's only used in compiler_error in case we need to emit the "too much static nesting" error.

File: Python/compile.c (hunk 3 above)
Comment (iritkatriel) [reply Python/compile.c]: I intended to remove these (in a separate commit, so I still have access in the future) but then forgot. They are places where I made a note that my changes didn't perfectly preserve the linenos as previously calculated. It's typically because the line number is not set there, so it was "inherited" from whatever happened just before.

File: Python/compile.c (hunk 2 above)
Comment (markshannon) [reply Python/compile.c]: We should eliminate dead code before assembly, so I don't think dead code is the problem.

Code that doesn't execute in lexical order might be. 
```
def f(x):
    return - (
        x)
```
Line 3 is the last line of code, but line 2 is the last executed line.

If this is a bug, it's not a new one, so don't worry about it for this PR.

**Review by markshannon — APPROVED** [review]
Thanks

File: Python/compile.c (hunk 2 above)
Comment (iritkatriel) [reply Python/compile.c]: > We should eliminate dead code before assembly, so I don't think dead code is the problem.

But we're sticking a line number on the reraise here, long before dead code elimination.

> If this is a bug, it's not a new one, so don't worry about it for this PR.

Indeed.

File: Python/compile.c (hunk 5 above)
Comment (iritkatriel) [reply Python/compile.c:4069]: I'll land this and do this and other cleanup in followup PRs. (There are still the SET_LOC and UNSET_LOC to remove - they do nothing now). 

Thanks for the review!
